batchflow/sampler.py

```diff
@@ -266,63 +198,174 @@ def truncate(self, high=None, low=None, expr=None, prob=0.5):
+
+class AndSampler(Sampler):
+    def __init__(self, left, right, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.left = left
+        self.right = right
+
+    def sample(self, size):
+        """ Sampling procedure of a product of two samplers.
+        """
+        _left = self.left.sample(size)
+        _right = self.right.sample(size)
+        return np.concatenate([_left, _right], axis=1)
+
+class ApplySampler(Sampler):
+    def __init__(self, sampler, transform, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.sampler = sampler
+        self.transform = transform
+
+    def sample(self, size):
+        """ Sampling procedure of a sampler subjugated to a transform.
+        """
+        return self.transform(self.sampler.sample(size))
+
+class TruncateSampler(Sampler):
+    def __init__(self, sampler, high=None, low=None, expr=None, prob=0.5, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.sampler = sampler
+        self.high = high
+        self.low = low
+        self.expr = expr
+        self.prob = prob
+
+    def sample(self, size):
+        """ Sampling method of a sampler subjugated to truncation.
+        """
+        if size == 0:
+            return self.sampler.sample(size=0)
+
+        high, low, expr, prob = self.high, self.low, self.expr, self.prob 
+        # set batch-size
+        expectation = size / prob
+        sigma = np.sqrt(size * (1 - prob) / (prob**2))
+        batch_size = int(expectation + 2 * sigma)
+
+        # sample, filter out, concat
+        ctr = 0
+        cumulated = 0
+        samples = []
+        while cumulated < size:
+            # sample points and compute condition-vector
+            sample = self.sampler.sample(size=batch_size)
+            cond = np.ones(shape=batch_size).astype(np.bool)
+            if low is not None:
+                if expr is not None:
+                    cond &= np.greater_equal(expr(sample).reshape(batch_size, -1), low).all(axis=1)
+                else:
+                    cond &= np.greater_equal(sample, low).all(axis=1)
+
+            if high is not None:
+                if expr is not None:
+                    cond &= np.less_equal(expr(sample).reshape(batch_size, -1), high).all(axis=1)
+                else:
+                    cond &= np.less_equal(sample, high).all(axis=1)
+
+            if high is None and low is None:
+                cond &= expr(sample).all(axis=1)
+
+            # check that truncation-prob is not to small
+            _share = np.sum(cond) / batch_size
+            if _share < SMALL_SHARE and ctr > 0:
+                raise ValueError('Probability of region of interest is too small. Try other truncation bounds')
+
+            # get points from region of interest
+            samples.append(sample[cond])
+            cumulated += np.sum(cond)
+            ctr += 1
+
+        return np.concatenate(samples)[:size]
+
+
+class BaseOperationSampler(Sampler):
```

Comment:
You can move it to `arithmetize` decorator, as well as all those two-lines classes

Reply:
I don't believe I can. If I do it like that (or in a similar manner):
```
def arithmetize(cls):
    class BaseOperationSampler(cls):
        ...
    class AddSampler(BaseOperationSampler):
        operation = '__add__'
        ...
    classes = ...
    for oper in ['__add__', ...]:
        def transform(self, other, fake=oper):
            _class = classes[fake]
            return _class(self, other)
        setattr(cls, oper, transform)

    return cls
```
`Sampler`-instances won't be _picklable_ anymore, as the classes are defined in a local namespace of `arithmetize`.
`Pool` will crash with `AttributeError: Can't pickle local object 'arithmetize.<locals>.AddSampler'`.

Check it out [here](http://82.148.26.146:8888/lab?) if interested.

Reply:
I believe that it is _possible_ by adding the names of dynamically generated classes to the module-scope, or by adding the `__reduce__` method to all of those classes; nevertheless, that is too much trouble